sw/airborne/modules/dcf/dcf.c

```diff
@@ -0,0 +1,178 @@
+ *
+ * This file is part of paparazzi.
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ */
+
+#include <math.h>
+#include <std.h>
+#include <stdio.h>
+
+#include "modules/dcf/dcf.h"
+#include "subsystems/datalink/datalink.h" // dl_buffer
+#include "subsystems/datalink/telemetry.h"
+#include "subsystems/navigation/common_nav.h"
+#include "autopilot.h"
+
+#if PERIODIC_TELEMETRY
+static void send_dcf(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_DCF(trans, dev, AC_ID, 4 * DCF_MAX_NEIGHBORS, &(tableNei[0][0]), DCF_MAX_NEIGHBORS, error_sigma);
+}
+#endif // PERIODIC TELEMETRY
+
+// Control
+/*! Default gain k for the algorithm */
+#ifndef DCF_GAIN_K
+#define DCF_GAIN_K 10
+#endif
+/*! Default radius for the circumference */
+#ifndef DCF_RADIUS
+#define DCF_RADIUS 80
+#endif
+/*! Default timeout for the neighbors' information */
+#ifndef DCF_TIMEOUT
+#define DCF_TIMEOUT 1500
+#endif
+
+dcf_con dcf_control = {DCF_GAIN_K, DCF_RADIUS, DCF_TIMEOUT, 0};
+
+/*! Default number of neighbors per aircraft */
+#ifndef DCF_MAX_NEIGHBORS
+#define DCF_MAX_NEIGHBORS 4
+#endif
+int16_t tableNei[DCF_MAX_NEIGHBORS][4];
+int16_t error_sigma[DCF_MAX_NEIGHBORS];
+uint32_t last_theta[DCF_MAX_NEIGHBORS];
+uint32_t last_transmision = 0;
+
+void dcf_init(void)
+{
+  for (int i = 0; i < DCF_MAX_NEIGHBORS; i++) {
+    tableNei[i][0] = -1;
+    error_sigma[i] = 0;
+  }
+
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_DCF, send_dcf);
+#endif
+}
+
+bool distributed_circular(uint8_t wp)
+{
+  float xc = waypoints[wp].x;
+  float yc = waypoints[wp].y;
+  struct EnuCoor_f *p = stateGetPositionEnu_f();
+  float x = p->x;
+  float y = p->y;
+  float u = 0;
+
+  dcf_control.theta = atan2f(y - yc, x - xc);
+
+  uint32_t now = get_sys_time_msec();
+
+  for (int i = 0; i < DCF_MAX_NEIGHBORS; i++) {
+    if (tableNei[i][0] != -1) {
+      uint32_t timeout = now - last_theta[i];
+      if (timeout > dcf_control.timeout) {
+        tableNei[i][3] = dcf_control.timeout;
+      } else {
+        tableNei[i][3] = (uint16_t)timeout;
+
+        float e = dcf_control.theta - (tableNei[i][1] + (tableNei[i][2])) * M_PI / 1800.0;
+        if (e > M_PI) {
+          e -= 2 * M_PI;
+        } else if (e <= -M_PI) {
+          e += 2 * M_PI;
+        }
```

Comment:
it is better to use a while loop to wrap the angle. You should use the macro NormRadAngle(e) (from include/std.h) here instead.

Reply:
The while loop is not necessary here, since by construction it is only required one iteration. Nevertheless, I will employ the suggested function. I suspected it existed :P .